scripts/get_objf_and_derivs_split.py

```diff
@@ -351,6 +357,57 @@ def MergeAndComputeObjfForSplit(split_index):
     ComputeObjfAndFinalDerivs(split_index, args.derivs_out != None)
 
+def CleanUpIfNeeded():
+    # For cleaning up intermediate files that won't be used, note if option need-model = true
+    # be used, clean up will keep file 'float.all'
+    if args.clean_up == 'true':
+        try:
+            if not os.path.isdir(args.work_dir):
+                ExitProgram("error finding working directory {0}".format(args.work_dir))
+            # If need_split_float option set to be true, this script will keep "float.all" in
+            # each split dir for future use
+            need_split_float = True if args.need_split_float == 'true' else False
+            for split_index in range(1, args.num_splits + 1):
+                split_dir = '{0}/split{1}/{2}'.format(args.work_dir,args.num_splits,split_index)
+                if not os.path.isdir(split_dir):
+                    ExitProgram("error finding split directory {0}".format(split_dir))
+                for file in os.listdir(split_dir):
+                    if re.match('(discounted|discounted_derivs|float|float_derivs|merged|merged_derivs)\.',file):
+
+                        if need_split_float and file == "float.all":
+                            continue
+                        os.remove(split_dir+"/"+file)
+                if need_split_float: continue
+                if os.listdir(split_dir) == []:
+                    os.rmdir(split_dir)
+                else:
+                    print("get_objf_and_drivs_split.py: split dir {0} in working dir has unexpected file,"\
+                          " check it if nesscessary".format(split_dir),file=sys.stderr)
+            if not need_split_float:
+                all_split_dir = '{0}/split{1}'.format(args.work_dir,args.num_splits)
+                if os.listdir(all_split_dir) == []:
+                    os.rmdir(all_split_dir)
+                else:
+                    print("get_objf_and_drivs_split.py: split dir {0} in working dir has unexpected file,"\
+                          " check it if nesscessary".format(all_split_dir),file=sys.stderr)
+            # If need_model option set to be true, this script will keep "float.all" in working dir
+            need_float = True if args.need_model == 'true' else False
+            for file in os.listdir(args.work_dir):
+                    if re.match('(discounted|discounted_derivs|float|float_derivs|merged|merged_derivs)\.',file):
+                        if need_float and file == "float.all":
+                            continue
+                        os.remove(args.work_dir+"/"+file)
+        except:
+            sys.exit("get_objf_and_drivs_split.py: error removing files in working dir")
+    try:
+        f = open(args.work_dir+'/cleaned', "w")
+        print(args.clean_up, file=f)
+        f.close()
+    except:
+        sys.exit("get_objf_and_derivs_split.py: error opening --cleaned={0} for writing".format(
+                  args.work_dir+'/cleaned'))
+
+
```

Comment:
the command get error is:

merge-float-counts data/lm_40k_3/float.all.1 data/lm_40k_3/float.all.2
data/lm_40k_3/float.all.3 data/lm_40k_3/float.all.4
data/lm_40k_3/float.all.5 >data/lm_40k_3_prune1.0/work/iter0/float.all


Running valgrind (after attach to debugger, it crashed with 'Failed to read
a valid object file image from memory.' )

valgrind --db-attach=yes  merge-float-counts data/lm_40k_3/float.all.1
data/lm_40k_3/float.all.2 data/lm_40k_3/float.all.3
data/lm_40k_3/float.all.4 data/lm_40k_3/float.all.5
>data/lm_40k_3_prune1.0/work/iter0/float.all

==5489== Memcheck, a memory error detector

==5489== Copyright (C) 2002-2011, and GNU GPL'd, by Julian Seward et al.

==5489== Using Valgrind-3.7.0 and LibVEX; rerun with -h for copyright info

==5489== Command: merge-float-counts data/lm_40k_3/float.all.1
data/lm_40k_3/float.all.2 data/lm_40k_3/float.all.3
data/lm_40k_3/float.all.4 data/lm_40k_3/float.all.5

==5489==

==5489== Invalid read of size 8

==5489==    at 0x402410: pocolm::FloatCountMerger::OutputState()
(merge-float-counts.cc:139)

==5489==    by 0x401899: pocolm::FloatCountMerger::FloatCountMerger(int,
char const**) (merge-float-counts.cc:51)

==5489==    by 0x401739: main (merge-float-counts.cc:184)

==5489==  Address 0x5aed7b8 is 8 bytes before a block of size 8 alloc'd

==5489==    at 0x4C286E7: operator new(unsigned long)
(vg_replace_malloc.c:287)

==5489==    by 0x407475: __gnu_cxx::new_allocator<long>::allocate(unsigned
long, void const*) (new_allocator.h:94)

==5489==    by 0x40676C: std::_Vector_base<long, std::allocator<long>
>::_M_allocate(unsigned long) (in
/export/a09/zhouyang/pocolm/src/merge-float-counts)

==5489==    by 0x4052BF: std::vector<long, std::allocator<long>
>::_M_fill_insert(__gnu_cxx::__normal_iterator<long*, std::vector<long,
std::allocator<long> > >, unsigned long, long const&) (vector.tcc:481)

==5489==    by 0x403C69: std::vector<long, std::allocator<long>
>::insert(__gnu_cxx::__normal_iterator<long*, std::vector<long,
std::allocator<long> > >, unsigned long, long const&) (stl_vector.h:1004)

==5489==    by 0x402D9E: std::vector<long, std::allocator<long>
>::resize(unsigned long, long) (stl_vector.h:687)

==5489==    by 0x40223A: pocolm::FloatCountMerger::OutputState()
(merge-float-counts.cc:118)

==5489==    by 0x401899: pocolm::FloatCountMerger::FloatCountMerger(int,
char const**) (merge-float-counts.cc:51)

==5489==    by 0x401739: main (merge-float-counts.cc:184)

==5489==

==5489==

==5489== ---- Attach to debugger ? --- [Return/N/n/Y/y/C/c] ---- y

==5489== starting debugger with cmd: /usr/bin/gdb -nw /proc/5723/fd/1024
5723

Failed to read a valid object file image from memory.

2016-07-27 18:14 GMT-04:00 Daniel Povey <notifications@github.com>:

> In scripts/get_objf_and_derivs_split.py
> <https://github.com/danpovey/pocolm/pull/30#discussion_r72534333>:
>
> > +            for file in os.listdir(args.work_dir):
> > +                    if re.match('(discounted|discounted_derivs|float|float_derivs|merged|merged_derivs)\.',file):
> > +                        if need_float and file == "float.all":
> > +                            continue
> > +                        os.remove(args.work_dir+"/"+file)
> > +        except:
> > +            sys.exit("get_objf_and_drivs_split.py: error removing files in working dir")
> > +    try:
> > +        f = open(args.work_dir+'/cleaned', "w")
> > +        print(args.clean_up, file=f)
> > +        f.close()
> > +    except:
> > +        sys.exit("get_objf_and_derivs_split.py: error opening --cleaned={0} for writing".format(
> > +                  args.work_dir+'/cleaned'))
> > +
> > +
>
> Regarding the double free error-- please figure out the full command line
> that is being run (the --verbose flag may help), and run it in valgrind,
> like valgrind --db-attach=yes program arg1 arg2.... then attach to debugger
> on the first error, go "up" until you find the first stack frame that's not
> in a system library, and show us what it's doing. Dan
> … <#m_4881438816306168021_>
> On Wed, Jul 27, 2016 at 2:58 PM, chris920820 ***@***.***> wrote: In
> scripts/get_objf_and_derivs_split.py <#30 (comment)
> <https://github.com/danpovey/pocolm/pull/30#discussion_r72532125>>: > +
> for file in os.listdir(args.work_dir): > + if
> re.match('(discounted|discounted_derivs|float|float_derivs|merged|merged_derivs)\.',file):
> > + if need_float and file == "float.all": > + continue > +
> os.remove(args.work_dir+"/"+file) > + except: > +
> sys.exit("get_objf_and_drivs_split.py: error removing files in working
> dir") > + try: > + f = open(args.work_dir+'/cleaned', "w") > +
> print(args.clean_up, file=f) > + f.close() > + except: > +
> sys.exit("get_objf_and_derivs_split.py: error opening --cleaned={0} for
> writing".format( > + args.work_dir+'/cleaned')) > + > + OK, I will do that.
> Plus, when I pull the version on your github master, tried to run examples
> (swbd and swbd_fisher), there seems to have some bug like this + for
> threshold in 1.0 0.25 + prune_lm_dir.py data/lm_20k_3 1.0
> data/lm_20k_3_prune1.0 + tail -n 5 + head -n 3 *** glibc detected ***
> merge-float-counts: double free or corruption (!prev): 0x00000000018570c0
> *** ======= Backtrace: =========
> /lib/x86_64-linux-gnu/libc.so.6(+0x75bb6)[0x7f7f17a79bb6]
> /lib/x86_64-linux-gnu/libc.so.6(cfree+0x6c)[0x7f7f17a7e95c]
> merge-float-counts[0x405fdc] merge-float-counts[0x4045c2]
> merge-float-counts[0x4053f1] merge-float-counts[0x403c6a]
> merge-float-counts[0x402d9f] merge-float-counts[0x40223b]
> merge-float-counts[0x40189a] merge-float-counts[0x40173a]
> /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xfd)[0x7f7f17a22ead]
> merge-float-counts[0x4015a9] ======= Memory map: ======== 00400000-00422000
> r-xp 00000000 00:24 11118193
> /export/a09/zhouyang/pocolm/src/merge-float-counts 00622000-00623000 rw-p
> 00022000 00:24 11118193 /export/a09/zhouyang/pocolm/src/merge-float-counts
> 01852000-01873000 rw-p 00000000 00:00 0 [heap] 7f7f10000000-7f7f10021000
> rw-p 00000000 00:00 0 7f7f10021000-7f7f14000000 ---p 00000000 00:00 0
> 7f7f17a04000-7f7f17b88000 r-xp 00000000 fe:00 75735 /lib/x86_64-linux-gnu/
> libc-2.13.so 7f7f17b88000-7f7f17d87000 ---p 00184000 fe:00 75735
> /lib/x86_64-linux-gnu/libc-2.13.so 7f7f17d87000-7f7f17d8b000 r--p
> 00183000 fe:00 75735 /lib/x86_64-linux-gnu/libc-2.13.so
> 7f7f17d8b000-7f7f17d8c000 rw-p 00187000 fe:00 75735 /lib/x86_64-linux-gnu/
> libc-2.13.so 7f7f17d8c000-7f7f17d91000 rw-p 00000000 00:00 0
> 7f7f17d91000-7f7f17da6000 r-xp 00000000 fe:00 49060
> /lib/x86_64-linux-gnu/libgcc_s.so.1 7f7f17da6000-7f7f17fa6000 ---p 00015000
> fe:00 49060 /lib/x86_64-linux-gnu/libgcc_s.so.1 7f7f17fa6000-7f7f17fa7000
> rw-p 00015000 fe:00 49060 /lib/x86_64-linux-gnu/libgcc_s.so.1
> 7f7f17fa7000-7f7f18028000 r-xp 00000000 fe:00 75739 /lib/x86_64-linux-gnu/
> libm-2.13.so 7f7f18028000-7f7f18227000 ---p 00081000 fe:00 75739
> /lib/x86_64-linux-gnu/libm-2.13.so 7f7f18227000-7f7f18228000 r--p
> 00080000 fe:00 75739 /lib/x86_64-linux-gnu/libm-2.13.so
> 7f7f18228000-7f7f18229000 rw-p 00081000 fe:00 75739 /lib/x86_64-linux-gnu/
> libm-2.13.so 7f7f18229000-7f7f18311000 r-xp 00000000 fe:00 1629156
> /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.17 7f7f18311000-7f7f18511000
> ---p 000e8000 fe:00 1629156 /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.17
> 7f7f18511000-7f7f18519000 r--p 000e8000 fe:00 1629156
> /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.17 7f7f18519000-7f7f1851b000
> rw-p 000f0000 fe:00 1629156 /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.17
> 7f7f1851b000-7f7f18530000 rw-p 00000000 00:00 0 7f7f18530000-7f7f18550000
> r-xp 00000000 fe:00 75732 /lib/x86_64-linux-gnu/ld-2.13.so
> 7f7f18665000-7f7f18732000 rw-p 00000000 00:00 0 7f7f1874b000-7f7f1874f000
> rw-p 00000000 00:00 0 7f7f1874f000-7f7f18750000 r--p 0001f000 fe:00 75732
> /lib/x86_64-linux-gnu/ld-2.13.so 7f7f18750000-7f7f18751000 rw-p 00020000
> fe:00 75732 /lib/x86_64-linux-gnu/ld-2.13.so 7f7f18751000-7f7f18752000
> rw-p 00000000 00:00 0 7ffcbac6a000-7ffcbac8b000 rw-p 00000000 00:00 0
> [stack] 7ffcbada9000-7ffcbadaa000 r-xp 00000000 00:00 0 [vdso]
> ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0 [vsyscall]
> 2016-07-27 16:23 GMT-04:00 Daniel Povey ***@***.***>: …
> <#m_-5486571734865106941_> In scripts/get_objf_and_derivs_split.py <#30
> <https://github.com/danpovey/pocolm/pull/30> (comment) <#30 (comment)
> <https://github.com/danpovey/pocolm/pull/30#discussion_r72515740>>>: > +
> for file in os.listdir(args.work_dir): > + if
> re.match('(discounted|discounted_derivs|float|float_derivs|merged|merged_derivs)\.',file):
> > + if need_float and file == "float.all": > + continue > +
> os.remove(args.work_dir+"/"+file) > + except: > +
> sys.exit("get_objf_and_drivs_split.py: error removing files in working
> dir") > + try: > + f = open(args.work_dir+'/cleaned', "w") > +
> print(args.clean_up, file=f) > + f.close() > + except: > +
> sys.exit("get_objf_and_derivs_split.py: error opening --cleaned={0} for
> writing".format( > + args.work_dir+'/cleaned')) > + > + FYI: for the
> top-level directory, I didn't want it to remove any files that have names
> not equal to those made by the script. People might put other stuff there
> for some reason and not expect them to be removed. What I was really hoping
> was that they could have the script just remove individually by name all
> the files it created, rather than messing about either with regexp's or
> searching through directories. This is safer, and not really that hard to
> implement. — You are receiving this because you authored the thread. Reply
> to this email directly, view it on GitHub <
> https://github.com/danpovey/pocolm/pull/30/files/9bf9b90520ae7bb207adafb9201d0718286763f5#r72515740>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/ASv4Rmtk9KxYUZ7WCPIW_wtVTmCrqiBaks5qZ75VgaJpZM4JI_LN>
> . — You are receiving this because you commented. Reply to this email
> directly, view it on GitHub <
> https://github.com/danpovey/pocolm/pull/30/files/9bf9b90520ae7bb207adafb9201d0718286763f5#r72532125>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/ADJVu9OcB7Hyx6c3FURVYugiOU0SJh_Xks5qZ9SHgaJpZM4JI_LN>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/pocolm/pull/30/files/9bf9b90520ae7bb207adafb9201d0718286763f5#r72534333>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASv4Rk-Hma3Z-WOdvV-lmWwvBbqmcRWUks5qZ9hKgaJpZM4JI_LN>
> .
>